kad-dht/README.md

```diff
@@ -0,0 +1,386 @@
+
+We keep track of:
+
+* the number of values we've fetched (`cnt`).
+* the best value we've found (`best`), and which peers returned it (`Pb`)
+* the set of peers we've already queried (`Pq`) and the set of next query candidates sorted by distance from `K` in ascending order (`Pn`).
+* the set of peers with outdated values (`Po`). 
+
+**Initialization**: seed `Pn` with the `α` peers from our routing table we know are closest to `K`, based on the XOR distance function.
+
+**Then we loop:**
+
+*WIP (raulk): lookup timeout.*
+
+1. If we have collected `Q` or more answers, we cancel outstanding requests, return `best`, and we notify the peers holding an outdated value (`Po`) of the best value we discovered, by sending `PUT_VALUE(K, best)` messages. _Return._
+2. Pick as many peers from the candidate peers (`Pn`) as the `α` concurrency factor allows. Send each a `GET_VALUE(K)` request, and mark it as _queried_ in `Pq`.
+3. Upon a response:
+	1. If successful, and we receive a value:
+		1. If this is the first value we've seen, we store it in `best`, along
+           with the peer who sent it in `Pb`.
+		2. Otherwise, we resolve the conflict by calling `Validator.Select(best,
+           new)`:
+			1. If the new value wins, store it in `best`, and mark all formerly
+               “best" peers (`Pb`) as _outdated peers_ (`Po`). The current peer
+               becomes the new best peer (`Pb`).
+			2. If the new value loses, we add the current peer to `Po`.
+	2. If successful without a value, the response will contain the closest
+       nodes the peer knows to the key `K`. Add them to the candidate list `Pn`,
+       except for those that have already been queried.
+	3. If an error or timeout occurs, discard it.
+4. Go to 1.
+
+## Entry validation
+
+When constructing a DHT node, it is possible to supply a record `Validator`
+object conforming to this interface:
+
+``` // Validator is an interface that should be implemented by record
+validators. type Validator interface {
+
+	// Validate validates the given record, returning an error if it's
+	// invalid (e.g., expired, signed by the wrong key, etc.).
+	Validate(key string, value []byte) error
+
+	// Select selects the best record from the set of records (e.g., the
+	// newest).
+	//
+	// Decisions made by select should be stable.
+	Select(key string, values [][]byte) (int, error)
+}
+```
+
+`Validate()` is a pure function that reports the validity of a record. It may
+validate a cryptographic signature, or else. It is called on two occasions:
+
+1. To validate incoming values in response to `GET_VALUE` calls.
+2. To validate outgoing values before storing them in the network via
+   `PUT_VALUE` calls.
+
+Similarly, `Select()` is a pure function that returns the best record out of 2
+or more candidates. It may use a sequence number, a timestamp, or other
+heuristic to make the decision.
```

Comment:
Rephrased in https://github.com/libp2p/specs/pull/108/commits/6909f8a97685c3f60679dbe7c9e81a5ab411bcf7. Happy to add more details on validation requirements, though I would prefer to merge a minimal version of the specification via this pull request first. Please comment on the pull request in case you object.